examples/twittermap/web/public/javascripts/mapresultcache/services.js

```diff
@@ -0,0 +1,82 @@
+/*
+ * Module to cache map result data. Common services module communicates with
+ * middleware and updates resultcache module.
+ */
+'use strict';
+angular.module('cloudberry.mapresultcache', ['cloudberry.common'])
+  .service('MapResultCache', function () {
+
+    // The key-value store that stores mapResults
+    var store = new HashMap();
+    // To check if keyword in query changed
+    var currentKeywords = [""];
+    // To check if time range in query changed
+    var date = new Date();
+    date.setDate(date.getDate() - 1);
+    var currentTimeRange = {
+      start: new Date(),
+      end: date
+    };
+    // Prefix for geoIds to make key unique
+    var prefix = Object.freeze({
+      state: 'S',
+      county: 'C',
+      city: 'I'
+    });
+    const INVALID_VALUE = 0;
+
+    // Check keyword, time range and the cache store
+    // Return the geoIds not in store
+    this.returnGeoIdsNotInCache = function (keywords, timeInterval, geoIds, geoLevel) {
+      // Length of geoIdsNotInCache is 0 in case of complete cache hit,
+      // same length as geoIds in case of complete cache miss,
+      // otherwise in range (0, geoIds.length)
+      var geoIdsNotInCache = [];
+
+      if(keywords.toString() != currentKeywords.toString() ||
+                !angular.equals(currentTimeRange, timeInterval)) {
+        store.clear();
+        currentKeywords = keywords.slice();
```

Comment:
why `slice`?

Reply:
It seems like the best way to shallow copy arrays in JavaScript. Regular assignment did not work.